python/mxnet/numpy/multiarray.py

```diff
@@ -720,6 +720,28 @@ def __getitem__(self, key):
                 key = (_np.newaxis,) + key
             return self._get_np_basic_indexing(key)
         elif indexing_dispatch_code == _NDARRAY_ADVANCED_INDEXING:
+            if prepend == _NDARRAY_NO_ZERO_DIM_BOOL_ARRAY and isinstance(key, (NDArray, _np.ndarray)):
```

Comment:
The `isinstance(key, (NDArray, _np.ndarray)` check here would be equivalent to the `isinstance(key, Symbol)` check in https://github.com/apache/incubator-mxnet/pull/18319/files#diff-e9b526dba36aa6f2e7dbb6c4d49e9822R159-R160

Can you elaborate why you call the `self._get_index_nd` here instead of directly calling `_npi.advanced_indexing(self, key)` as in the Symbol case?

Reply:
I think this is for boolean indexing.

Reply:
So boolean indexing is now supported by `_npi.advanced_indexing`?

Reply:
I think this is to tell whether there is 0-dim/boolean indexing in the key.

Reply:
@leezu https://github.com/apache/incubator-mxnet/blob/a9b16f7024878611b236c9f3734ccd37a5a35d38/python/mxnet/ndarray/ndarray.py#L116-L119

Reply:
Yes. My question is that it appears that `indexing_key_expand_implicit_axes` messes up the `key` as it supports a much more generic case. It may be easier to have the `if` condition and call to `_npi.advanced_indexing` prior to the `indexing_key_expand_implicit_axes`. In that case the code may be simply

```
if isinstance(key, (NDArray, _np.ndarray) and key.dtype is not _np.bool:
    _npi.advanced_indexing(self, key)
```

But maybe I'm missing something, so clarification is appreciated. Thanks

Reply:
Thanks for reviewing my code @leezu ! Actually for now my op does not cover newaxis indexing case in backend since the primary purpose of this PR is to solve the issue #17327 (support indexing like sequence[indices_x, indices_y] after hybridized). This is why this if statement is so complex.

Reply:
@AntiZpvoh Thank you for clarifying! However I'm still not clear why the ndarray codepath needs to be more complex than the symbol one for the case that you ensure that the key is either a single ndarray or a tuple of ndarrays. These are the two cases supported in the symbol interface, so you don't need to worry about other cases in the imperative interface.

Reply:
@leezu My apology I found my mistake. It should be `isinstance(key, tuple)` instead of `isinstance(key, (NDArray, _np.ndarray))`. In that case codepath needs to be complex since we need to ensure the element of tuple is integer NDArray or np.ndarray (cannot be list maybe because list could not be symbolized) and also handle np.newaxis and implicit axis cases. The newest commit has already corrected this mistake.

Reply:
> In that case codepath needs to be complex since we need to ensure the element of tuple is integer NDArray or np.ndarray (cannot be list maybe because list could not be symbolized) and also handle np.newaxis and implicit axis cases. The newest commit has already corrected this mistake.

Thanks for fixing the mistake. My concern is that you need to handle the special cases "tuple of NDArray" without `_get_index_nd` and even without `indexing_key_expand_implicit_axes`. Consider your current test file `test_gluon_indexing.py`. It will pass the `key = (array([0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7], dtype=int32), array([27, 30, 24, 26,  1, 23, 12, 25, 29, 22, 21, 31,  6, 31, 11,  6], dtype=int32))`, but after you call `indexing_key_expand_implicit_axes` the key is transformed to `(array([0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7], dtype=int32), array([27, 30, 24, 26,  1, 23, 12, 25, 29, 22, 21, 31,  6, 31, 11,  6], dtype=int32), slice(None, None, None))` and your condition `if is_int_array_tuple` will never be True and `_npi.advanced_indexing_multiple` will never be called..